app/driver.go

```diff
@@ -370,6 +374,115 @@ func getFlexvolumeDaemonSetSpec(image, flexvolumeDir string) *appsv1beta2.Daemon
+	flexvolumeDir := ""
+
+	// start server
+	doneChan := make(chan struct{})
+	s := &http.Server{
+		Addr: fmt.Sprintf(":%d", serverPort),
+	}
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		logrus.Infof("Discovered FlexVolume dir path: %s", r.URL.Path)
+		flexvolumeDir = r.URL.Path
+		close(doneChan)
+	})
+	go s.ListenAndServe()
+
+	// create self-targeting service
+	objectMeta := metav1.ObjectMeta{
+		Name: "discover-flexvolume-dir",
+	}
+	kubeClient.CoreV1().Services(namespace).Create(&v1.Service{
+		ObjectMeta: objectMeta,
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				v1.ServicePort{
+					Name: "longhorn-driver-deployer",
+					Port: serverPort,
+				},
+			},
+			Selector: map[string]string{
+				"app": "longhorn-driver-deployer",
+			},
+			// FIXME
+			// Type: v1.ServiceTypeClusterIP,
+			Type: v1.ServiceTypeNodePort,
+		},
+	})
+
+	// create job that will send results to our server
+	privilege := true
+	_, err = kubeClient.BatchV1().Jobs(namespace).Create(&batchv1.Job{
+		ObjectMeta: objectMeta,
+		Spec: batchv1.JobSpec{
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: objectMeta,
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						v1.Container{
+							Name:    objectMeta.Name,
+							Image:   "busybox",
+							Command: []string{"/bin/sh"},
+							Args: []string{"-c",
+								`find_kubelet_proc() {
+					        for proc in $(find /proc -type d -maxdepth 1); do
+					          if [ ! -f $proc/cmdline ]; then
+					            continue
+					          fi
+					          if [[ "$(cat $proc/cmdline | tr '\000' '\n' | head -n1 | tr '/' '\n' | tail -n1)" == "kubelet" ]]; then
+					            echo $proc
+					            return
+					          fi
+					        done
+					      }
+					      get_flexvolume_path() {
+					        proc=$(find_kubelet_proc)
+					        if [ "$proc" != "" ]; then
+					          path=$(cat $proc/cmdline | tr '\000' '\n' | grep volume-plugin-dir | tr '=' '\n' | tail -n1)
+					          if [ "$path" == "" ]; then
+					            echo '/usr/libexec/kubernetes/kubelet-plugins/volume/exec/'
+					          else
+					            echo $path
+					          fi
+					          return
+					        fi
+					        echo 'no kubelet process found, dunno'
+					      }
+					      FLEXVOLUME_PATH=$(get_flexvolume_path)
+					      wget -q -O - ${DISCOVER_FLEXVOLUME_DIR_SERVICE_HOST}:${DISCOVER_FLEXVOLUME_DIR_SERVICE_PORT}/${FLEXVOLUME_PATH}
+					      `},
+							SecurityContext: &v1.SecurityContext{
+								Privileged: &privilege,
+							},
+						},
+					},
+					RestartPolicy: v1.RestartPolicyNever,
+					HostPID:       true,
+				},
+			},
+		},
+	})
+	if err != nil {
+		// cleanup async
+		return "", err
+	}
+
+	<-doneChan
+	ctx, cancel := context.WithTimeout(nil, 1*time.Second)
+	s.Shutdown(ctx)
+	cancel()
+
+	// cleanup async
+
+	return flexvolumeDir, nil
+}
+
 func discoverFlexvolumeDir(kubeClient *clientset.Clientset) (dir string, err error) {
```

Comment:
@yasker What should we do with the old method of detection? Just delete it?

Reply:
The old way is much faster. Check the version and if it's <= 1.8, run the old way first. If the old way cannot get the result, run the new way.

Reply:
In fact, thinking about it again, only 1.8 can use it. Not worth the effort to keep it. Let's remove it and switch altogether.